File: backend/src-tauri/src/commands/timeline.rs
use crate::state::AppState;
use base64::Engine;
use std::fs;
use std::path::Path;
use tauri::Emitter;
use tauri::Manager;
use tauri::State;
use uuid::Uuid;

use super::common::ensure_temp_dir;

#[derive(Debug, Clone, serde::Serialize)]
#[serde(rename_all = "camelCase")]
struct ClipFormantStatusPayload {
    clip_id: String,
    status: String,
}

fn emit_clip_formant_status(app: &tauri::AppHandle, clip_id: &str, status: &str) {
    let _ = app.emit(
        "clip_formant_status",
        ClipFormantStatusPayload {
            clip_id: clip_id.to_string(),
            status: status.to_string(),
        },
    );
}

fn clip_formant_rebuild_needs_refresh(
    before: Option<&crate::state::Clip>,
    after: &crate::state::Clip,
) -> bool {
    let before_enabled = before
        .and_then(|clip| clip.formant_morph.as_ref())
        .map(|params| params.enabled)
        .unwrap_or(false);
    let after_enabled = after
        .formant_morph
        .as_ref()
        .map(|params| params.enabled)
        .unwrap_or(false);

    if !before_enabled && !after_enabled {
        return false;
    }

    let before_params = before.and_then(|clip| clip.formant_morph.as_ref());
    let after_params = after.formant_morph.as_ref();
    let params_changed = before_params != after_params;
    let source_changed = before
        .map(|clip| {
            clip.source_path != after.source_path
                || (clip.source_start_sec - after.source_start_sec).abs() > 1e-9
                || (clip.source_end_sec - after.source_end_sec).abs() > 1e-9
                || clip.reversed != after.reversed
        })
        .unwrap_or(after_enabled);

    params_changed || source_changed
}

fn schedule_clip_formant_rebuild(state: &AppState, clip: crate::state::Clip) {
    let Some(app) = state.app_handle.get().cloned() else {
        return;
    };
    let clip_id = clip.id.clone();
    let Some(formant) = clip.formant_morph.as_ref() else {
        crate::formant_cache::cancel_formant_rebuild_generation(&clip_id);
        return;
    };
    if !formant.enabled {
        crate::formant_cache::cancel_formant_rebuild_generation(&clip_id);
        return;
    }

    let generation = crate::formant_cache::begin_formant_rebuild_generation(&clip_id);
    if let Some(formant) = clip.formant_morph.as_ref() {
        crate::formant_cache::formant_debug_log(format!(
            "schedule rebuild clip_id={} generation={} f1={:.1} f2={:.1} strength={:.3} source={} range=[{:.3},{:.3}] reversed={}",
            clip_id,
            generation,
            formant.target_f1_hz,
            formant.target_f2_hz,
            formant.strength,
            clip.source_path.as_deref().unwrap_or(""),
            clip.source_start_sec,
            clip.source_end_sec,
            clip.reversed,
        ));
    }
    emit_clip_formant_status(&app, &clip_id, "rebuilding");

    let out_rate = state.audio_engine.sample_rate_hz().max(8_000);
    std::thread::spawn(move || {
        let result = crate::formant_cache::compute_formant_cache_entry_for_clip(&clip, out_rate);

        if !crate::formant_cache::is_current_formant_rebuild_generation(&clip_id, generation) {
            crate::formant_cache::formant_debug_log(format!(
                "discard stale rebuild clip_id={} generation={}",
                clip_id, generation
            ));
            return;
        }

        match result {
            Ok((key, entry)) => {
                crate::formant_cache::formant_debug_log(format!(
                    "rebuild ready clip_id={} generation={} frames={} sr={}",
                    clip_id, generation, entry.frames, entry.sample_rate
                ));
                crate::formant_cache::insert_formant_cache_entry(key, entry);
                emit_clip_formant_status(&app, &clip_id, "ready");
                let state = app.state::<AppState>();
                let timeline = match state.timeline.lock() {
                    Ok(guard) => guard.clone(),
                    Err(poisoned) => poisoned.into_inner().clone(),
                };
                state.audio_engine.update_timeline(timeline);
            }
            Err(error) => {
                crate::formant_cache::formant_debug_log(format!(
                    "rebuild failed clip_id={} generation={} error={}",
                    clip_id, generation, error
                ));
                emit_clip_formant_status(&app, &clip_id, "failed");
            }
        }
    });
}

// ===================== dialogs / io =====================

pub(super) fn import_audio_bytes(
    state: State<'_, AppState>,
    file_name: String,
    base64_data: String,
    track_id: Option<Option<String>>,
    start_sec: Option<f64>,
) -> crate::models::TimelineStatePayload {
    if std::env::var("HIFISHIFTER_DEBUG_COMMANDS").ok().as_deref() == Some("1") {
        eprintln!(
            "import_audio_bytes(file_name={}, base64_len={}, track_id={:?}, start_sec={:?})",
            file_name,
            base64_data.len(),
            track_id,
            start_sec
        );
    }
    let engine = base64::engine::general_purpose::STANDARD;
    let bytes = engine.decode(base64_data.as_bytes()).unwrap_or_default();

    let ext = Path::new(&file_name)
        .extension()
        .and_then(|e| e.to_str())
        .unwrap_or("bin");
    let tmp_dir = ensure_temp_dir().ok();
    let path = tmp_dir.unwrap_or_else(std::env::temp_dir).join(format!(
        "{}_{}.{}",
        "import",
        Uuid::new_v4().simple(),
        ext
    ));

    let _ = fs::write(&path, &bytes);

    let mut tl = state.timeline.lock().unwrap_or_else(|e| e.into_inner());
    state.checkpoint_timeline(&tl);
    let resolved_track_id: Option<String> = match track_id {
        None => None,
        Some(Some(id)) => Some(id),
        Some(None) => Some(tl.add_track(Some("Track".to_string()), None, None)),
    };

    tl.import_audio_item(&path.display().to_string(), resolved_track_id, start_sec);
    state.audio_engine.update_timeline(tl.clone());
    let mut payload = tl.to_payload();
    payload.project = Some(state.project_meta_payload());
    payload
}

pub(super) fn import_audio_item(
    state: State<'_, AppState>,
    audio_path: String,
    track_id: Option<Option<String>>,
    start_sec: Option<f64>,
) -> crate::models::TimelineStatePayload {
    if std::env::var("HIFISHIFTER_DEBUG_COMMANDS").ok().as_deref() == Some("1") {
        eprintln!(
            "import_audio_item(audio_path={}, track_id={:?}, start_sec={:?})",
            audio_path, track_id, start_sec
        );
    }
    {
        let mut rt = state.runtime.lock().unwrap_or_else(|e| e.into_inner());
        rt.audio_loaded = true;
    }

    let mut tl = state.timeline.lock().unwrap_or_else(|e| e.into_inner());
    state.checkpoint_timeline(&tl);
    let resolved_track_id: Option<String> = match track_id {
        None => None,
        Some(Some(id)) => Some(id),
        Some(None) => Some(tl.add_track(Some("Track".to_string()), None, None)),
    };

    tl.import_audio_item(&audio_path, resolved_track_id, start_sec);
    state.audio_engine.update_timeline(tl.clone());
    let mut payload = tl.to_payload();
    payload.project = Some(state.project_meta_payload());
    payload
}

// ===================== timeline CRUD =====================

pub(super) fn add_track(
    state: State<'_, AppState>,
    name: Option<String>,
    parent_track_id: Option<String>,
    index: Option<usize>,
) -> crate::models::TimelineStatePayload {
    let mut tl = state.timeline.lock().unwrap_or_else(|e| e.into_inner());
    state.checkpoint_timeline(&tl);
    tl.add_track(name, parent_track_id, index);
    state.audio_engine.update_timeline(tl.clone());
    let mut payload = tl.to_payload();
    payload.project = Some(state.project_meta_payload());
    payload
}

pub(super) fn remove_track(
    state: State<'_, AppState>,
    track_id: String,
) -> crate::models::TimelineStatePayload {
    let mut tl = state.timeline.lock().unwrap_or_else(|e| e.into_inner());
    state.checkpoint_timeline(&tl);

    // 删除前：BFS 收集将被删除的轨道 ID 及其关联的 clip ID，用于后续清理全局缓存。
    let (clip_ids_to_clean, root_track_ids_to_clean) = {
        let mut to_remove = vec![track_id.clone()];
        let mut idx = 0;
        while idx < to_remove.len() {
            let cur = to_remove[idx].clone();
            for child in tl
                .tracks
                .iter()
                .filter(|t| t.parent_id.as_deref() == Some(cur.as_str()))
                .map(|t| t.id.clone())
            {
                to_remove.push(child);
            }
            idx += 1;
        }
        let remove_set: std::collections::HashSet<&str> =
            to_remove.iter().map(|s| s.as_str()).collect();
        let clip_ids: Vec<String> = tl
            .clips
            .iter()
            .filter(|c| remove_set.contains(c.track_id.as_str()))
            .map(|c| c.id.clone())
            .collect();
        (clip_ids, to_remove)
    };

    tl.remove_track(&track_id);
    state.audio_engine.update_timeline(tl.clone());

    // 清理被删除 clip 的全局合成缓存和渲染状态，防止内存泄漏和旧数据残留。
    for clip_id in &clip_ids_to_clean {
        crate::synth_clip_cache::invalidate_clip_all_caches(clip_id);
    }

    // 将锁的获取移到循环外部，避免 O(N) 的锁争用开销
    if let Ok(mut mgr) = crate::clip_rendering_state::global_clip_rendering_state().lock() {
        for clip_id in &clip_ids_to_clean {
            mgr.remove_state(clip_id);
        }
    }

    // 清理被删除轨道的 pitch_timeline_snapshot，防止增量分析数据残留。
    if let Ok(mut snapshot_map) = state.pitch_timeline_snapshot.lock() {
        for root_id in &root_track_ids_to_clean {
            snapshot_map.remove(root_id);
        }
    }

    // 清理 pitch_inflight 中包含被删轨道 ID 的去重 key。
    if let Ok(mut inflight) = state.pitch_inflight.lock() {
        inflight.retain(|key| {
            !root_track_ids_to_clean
                .iter()
                .any(|tid| key.contains(tid.as_str()))
        });
    }

    let mut payload = tl.to_payload();
    payload.project = Some(state.project_meta_payload());
    payload
}

pub(super) fn duplicate_track(
    state: State<'_, AppState>,
    track_id: String,
) -> crate::models::TimelineStatePayload {
    let mut tl = state.timeline.lock().unwrap_or_else(|e| e.into_inner());
    state.checkpoint_timeline(&tl);
    tl.duplicate_track(&track_id);
    state.audio_engine.update_timeline(tl.clone());
    let mut payload = tl.to_payload();
    payload.project = Some(state.project_meta_payload());
    payload
}

pub(super) fn move_track(
    state: State<'_, AppState>,
    track_id: String,
    target_index: usize,
    parent_track_id: Option<String>,
) -> crate::models::TimelineStatePayload {
    let mut tl = state.timeline.lock().unwrap_or_else(|e| e.into_inner());
    state.checkpoint_timeline(&tl);
    tl.move_track(&track_id, target_index, parent_track_id);
    state.audio_engine.update_timeline(tl.clone());
    let mut payload = tl.to_payload();
    payload.project = Some(state.project_meta_payload());
    payload
}

pub(super) fn set_track_state(
    state: State<'_, AppState>,
    track_id: String,
    muted: Option<bool>,
    solo: Option<bool>,
    volume: Option<f32>,
    compose_enabled: Option<bool>,
    pitch_analysis_algo: Option<String>,
    color: Option<String>,
    name: Option<String>,
) -> crate::models::TimelineStatePayload {
    let mut tl = state.timeline.lock().unwrap_or_else(|e| e.into_inner());
    state.checkpoint_timeline(&tl);
    let algo = pitch_analysis_algo.as_deref().map(|s| match s {
        "world_dll" | "world" => crate::state::PitchAnalysisAlgo::WorldDll,
        "nsf_hifigan_onnx" | "nsf_hifigan" | "onnx" => {
            crate::state::PitchAnalysisAlgo::NsfHifiganOnnx
        }
        "vslib" | "vocalshifter_vslib" => crate::state::PitchAnalysisAlgo::VocalShifterVslib,
        "none" => crate::state::PitchAnalysisAlgo::None,
        _ => crate::state::PitchAnalysisAlgo::Unknown,
    });
    tl.set_track_state(
        &track_id,
        muted,
        solo,
        volume,
        compose_enabled,
        algo,
        color,
        name,
    );
    state.audio_engine.update_timeline(tl.clone());
    let mut payload = tl.to_payload();
    payload.project = Some(state.project_meta_payload());
    payload
}

pub(super) fn select_track(
    state: State<'_, AppState>,
    track_id: String,
) -> crate::models::TimelineStatePayload {
    let mut tl = state.timeline.lock().unwrap_or_else(|e| e.into_inner());
    tl.select_track(&track_id);
    state.audio_engine.update_timeline(tl.clone());
    let mut payload = tl.to_payload();
    payload.project = Some(state.project_meta_payload());
    payload
}

pub(super) fn set_project_length(
    state: State<'_, AppState>,
    project_sec: f64,
) -> crate::models::TimelineStatePayload {
    let mut tl = state.timeline.lock().unwrap_or_else(|e| e.into_inner());
    state.checkpoint_timeline(&tl);
    tl.set_project_length(project_sec);
    state.audio_engine.update_timeline(tl.clone());
    let mut payload = tl.to_payload();
    payload.project = Some(state.project_meta_payload());
    payload
}

pub(super) fn add_clip(
    state: State<'_, AppState>,
    track_id: Option<String>,
    name: Option<String>,
    start_sec: Option<f64>,
    length_sec: Option<f64>,
    source_path: Option<String>,
) -> crate::models::TimelineStatePayload {
    let mut tl = state.timeline.lock().unwrap_or_else(|e| e.into_inner());
    state.checkpoint_timeline(&tl);
    tl.add_clip(track_id, name, start_sec, length_sec, source_path);
    state.audio_engine.update_timeline(tl.clone());
    let mut payload = tl.to_payload();
    payload.project = Some(state.project_meta_payload());
    payload
}

pub(super) fn create_clips_bulk(
    state: State<'_, AppState>,
    payload: crate::state::CreateClipsBulkPayload,
) -> crate::models::TimelineStatePayload {
    let mut tl = state.timeline.lock().unwrap_or_else(|e| e.into_inner());
    state.checkpoint_timeline(&tl);
    let created_clip_ids = tl.create_clips_bulk(&payload);
    state.audio_engine.update_timeline(tl.clone());
    let mut timeline_payload = tl.to_payload();
    timeline_payload.created_clip_ids = Some(created_clip_ids);
    timeline_payload.project = Some(state.project_meta_payload());
    timeline_payload
}

pub(super) fn remove_clip(
    state: State<'_, AppState>,
    clip_id: String,
) -> crate::models::TimelineStatePayload {
    crate::formant_cache::cancel_formant_rebuild_generation(&clip_id);
    let mut tl = state.timeline.lock().unwrap_or_else(|e| e.into_inner());
    state.checkpoint_timeline(&tl);
    tl.remove_clip(&clip_id);
    state.audio_engine.update_timeline(tl.clone());
    let mut payload = tl.to_payload();
    payload.project = Some(state.project_meta_payload());
    payload
}

pub(super) fn remove_clips(
    state: State<'_, AppState>,
    clip_ids: Vec<String>,
) -> crate::models::TimelineStatePayload {
    for clip_id in &clip_ids {
        crate::formant_cache::cancel_formant_rebuild_generation(clip_id);
    }
    let mut tl = state.timeline.lock().unwrap_or_else(|e| e.into_inner());
    state.checkpoint_timeline(&tl);
    tl.remove_clips(&clip_ids);
    state.audio_engine.update_timeline(tl.clone());
    let mut payload = tl.to_payload();
    payload.project = Some(state.project_meta_payload());
    payload
}

pub(super) fn move_clip(
    state: State<'_, AppState>,
    clip_id: String,
    start_sec: f64,
    track_id: Option<String>,
    move_linked_params: Option<bool>,
) -> crate::models::TimelineStatePayload {
    let mut tl = state.timeline.lock().unwrap_or_else(|e| e.into_inner());
    state.checkpoint_timeline(&tl);
    tl.move_clip(
        &clip_id,
        start_sec,
        track_id,
        move_linked_params.unwrap_or(false),
    );
    state.audio_engine.update_timeline(tl.clone());
    let mut payload = tl.to_payload();
    payload.project = Some(state.project_meta_payload());
    payload
}

pub(super) fn move_clips(
    state: State<'_, AppState>,
    moves: Vec<crate::state::MoveClipPayload>,
    move_linked_params: Option<bool>,
) -> crate::models::TimelineStatePayload {
    let mut tl = state.timeline.lock().unwrap_or_else(|e| e.into_inner());
    state.checkpoint_timeline(&tl);
    tl.move_clips(&moves, move_linked_params.unwrap_or(false));
    state.audio_engine.update_timeline(tl.clone());
    let mut payload = tl.to_payload();
    payload.project = Some(state.project_meta_payload());
    payload
}

pub(super) fn get_clip_linked_params(
    state: State<'_, AppState>,
    clip_id: String,
) -> serde_json::Value {
    let mut tl = state.timeline.lock().unwrap_or_else(|e| e.into_inner());
    match tl.extract_clip_linked_params(&clip_id) {
        Some(linked_params) => serde_json::json!({
            "ok": true,
            "linkedParams": linked_params,
        }),
        None => serde_json::json!({
            "ok": false,
            "error": "clip_not_found",
        }),
    }
}

pub(super) fn apply_clip_linked_params(
    state: State<'_, AppState>,
    clip_id: String,
    linked_params: crate::state::LinkedParamCurvesPayload,
) -> crate::models::TimelineStatePayload {
    let mut tl = state.timeline.lock().unwrap_or_else(|e| e.into_inner());
    state.checkpoint_timeline(&tl);
    tl.apply_linked_params_to_clip(&clip_id, &linked_params);
    state.audio_engine.update_timeline(tl.clone());
    let mut payload = tl.to_payload();
    payload.project = Some(state.project_meta_payload());
    payload
}

#[allow(clippy::too_many_arguments)]

pub(super) fn set_clip_state(
    state: State<'_, AppState>,
    clip_id: String,
    name: Option<String>,
    start_sec: Option<f64>,
    length_sec: Option<f64>,
    gain: Option<f32>,
    muted: Option<bool>,
    source_start_sec: Option<f64>,
    source_end_sec: Option<f64>,
    playback_rate: Option<f32>,
    reversed: Option<bool>,
    fade_in_sec: Option<f64>,
    fade_out_sec: Option<f64>,
    fade_in_curve: Option<String>,
    fade_out_curve: Option<String>,
    color: Option<String>,
    formant_morph: Option<crate::state::ClipFormantMorph>,
    checkpoint: Option<bool>,
) -> crate::models::TimelineStatePayload {
    let mut tl = state.timeline.lock().unwrap_or_else(|e| e.into_inner());
    let previous_clip = tl.clips.iter().find(|clip| clip.id == clip_id).cloned();
    // checkpoint 默认为 true，但可以通过传递 false 来抑制 undo checkpoint
    // 这在 undo group 内进行多次操作时很有用
    let do_checkpoint = checkpoint.unwrap_or(true);
    if do_checkpoint {
        state.checkpoint_timeline(&tl);
    }
    tl.patch_clip_state(
        &clip_id,
        crate::state::ClipStatePatch {
            name,
            start_sec,
            length_sec,
            gain,
            muted,
            source_start_sec,
            source_end_sec,
            playback_rate,
            reversed,
            fade_in_sec,
            fade_out_sec,
            fade_in_curve,
            fade_out_curve,
            color,
            formant_morph,
        },
    );
    let next_clip = tl.clips.iter().find(|clip| clip.id == clip_id).cloned();
    state.audio_engine.update_timeline(tl.clone());
    let mut payload = tl.to_payload();
    payload.project = Some(state.project_meta_payload());
    drop(tl);

    if let Some(next_clip) = next_clip {
        if clip_formant_rebuild_needs_refresh(previous_clip.as_ref(), &next_clip) {
            schedule_clip_formant_rebuild(&state, next_clip);
        }
    }
    payload
}

pub(super) fn set_clips_state_bulk(
    state: State<'_, AppState>,
    updates: Vec<crate::state::BulkClipStatePatch>,
    checkpoint: Option<bool>,
) -> crate::models::TimelineStatePayload {
    let mut tl = state.timeline.lock().unwrap_or_else(|e| e.into_inner());
    if checkpoint.unwrap_or(true) {
        state.checkpoint_timeline(&tl);
    }
    tl.patch_clips_state(&updates);
    state.audio_engine.update_timeline(tl.clone());
    let mut payload = tl.to_payload();
    payload.project = Some(state.project_meta_payload());
    payload
}

pub(super) fn duplicate_clips_bulk(
    state: State<'_, AppState>,
    payload: crate::state::DuplicateClipsBulkPayload,
) -> crate::models::TimelineStatePayload {
    let mut tl = state.timeline.lock().unwrap_or_else(|e| e.into_inner());
    state.checkpoint_timeline(&tl);
    let created_clip_ids = tl.duplicate_clips_bulk(&payload);
    state.audio_engine.update_timeline(tl.clone());
    let mut timeline_payload = tl.to_payload();
    timeline_payload.created_clip_ids = Some(created_clip_ids);
    timeline_payload.project = Some(state.project_meta_payload());
    timeline_payload
}

pub(super) fn replace_clip_source(
    state: State<'_, AppState>,
    clip_ids: Vec<String>,
    new_source_path: String,
    replace_same_source: Option<bool>,
) -> crate::models::TimelineStatePayload {
    let mut tl = state.timeline.lock().unwrap_or_else(|e| e.into_inner());
    state.checkpoint_timeline(&tl);
    tl.replace_clip_sources(
        &clip_ids,
        &new_source_path,
        replace_same_source.unwrap_or(false),
    );
    state.audio_engine.update_timeline(tl.clone());
    let mut payload = tl.to_payload();
    payload.project = Some(state.project_meta_payload());
    payload
}

pub(super) fn split_clip(
    state: State<'_, AppState>,
    clip_id: String,
    split_sec: f64,
) -> crate::models::TimelineStatePayload {
    let mut tl = state.timeline.lock().unwrap_or_else(|e| e.into_inner());
    state.checkpoint_timeline(&tl);
    tl.split_clip(&clip_id, split_sec);
    state.audio_engine.update_timeline(tl.clone());
    let mut payload = tl.to_payload();
    payload.project = Some(state.project_meta_payload());
    payload
}

pub(super) fn glue_clips(
    state: State<'_, AppState>,
    clip_ids: Vec<String>,
) -> crate::models::TimelineStatePayload {
    let mut tl = state.timeline.lock().unwrap_or_else(|e| e.into_inner());
    state.checkpoint_timeline(&tl);
    tl.glue_clips(&clip_ids);
    state.audio_engine.update_timeline(tl.clone());
    let mut payload = tl.to_payload();
    payload.project = Some(state.project_meta_payload());
    payload
}

pub(super) fn select_clip(
    state: State<'_, AppState>,
    clip_id: Option<String>,
) -> crate::models::TimelineStatePayload {
    let mut tl = state.timeline.lock().unwrap_or_else(|e| e.into_inner());
    tl.select_clip(clip_id);
    let mut payload = tl.to_payload();
    payload.project = Some(state.project_meta_payload());
    payload
}

pub(super) fn get_track_summary(
    state: State<'_, AppState>,
    track_id: Option<String>,
) -> serde_json::Value {
    // Minimal placeholder summary; waveform is empty until audio pipeline is migrated.
    let tl = state.timeline.lock().unwrap_or_else(|e| e.into_inner());
    let tid = track_id
        .or_else(|| tl.selected_track_id.clone())
        .or_else(|| tl.tracks.first().map(|t| t.id.clone()))
        .unwrap_or_default();

    let clip_count = tl.clips.iter().filter(|c| c.track_id == tid).count();

    serde_json::json!({
        "ok": true,
        "track_id": tid,
        "clip_count": clip_count,
        "waveform_preview": [],
        "pitch_range": {"min": -24, "max": 24}
    })
}

File: frontend/src/services/api/timeline.ts
import type { TimelineResult, TrackSummaryResult } from "../../types/api";
import type { LinkedParamCurves } from "../../features/session/sessionTypes";

import { invoke } from "../invoke";
import type { ClipTemplate } from "../../features/session/sessionTypes";

export interface SampleRegionAnnotation {
    name: string;
    region_start_sec: number;
    region_end_sec: number;
    note_alignment_sec: number;
    fixed_duration_sec: number;
}

export interface SamplePitchNote {
    start_sec: number;
    end_sec: number;
    midi_note: number;
    confidence: number;
}

export type ClipSampleAnnotationsResult =
    | {
          ok: true;
          clip_id: string;
          audio_path: string;
          sidecar_path: string;
          annotations: SampleRegionAnnotation[];
          pitch_notes: SamplePitchNote[];
          active_annotation_index: number;
      }
    | { ok: false; error: string };

export type SaveClipSampleAnnotationsResult =
    | {
          ok: true;
          clip_id: string;
          sidecar_path: string;
          annotations: SampleRegionAnnotation[];
          active_annotation_index: number;
      }
    | { ok: false; error: string };

export interface OtoConversionResult {
    oto_files: number;
    converted_samples: Array<{
        audio_path: string;
        sidecar_path: string;
        annotation_count: number;
    }>;
    warnings: string[];
}

export const timelineApi = {
    // Undo/Redo (backend-authoritative)
    undoTimeline: () => invoke<TimelineResult>("undo_timeline"),
    redoTimeline: () => invoke<TimelineResult>("redo_timeline"),

    // Undo grouping: all commands between begin/end share a single undo entry
    beginUndoGroup: () => invoke<TimelineResult>("begin_undo_group"),
    endUndoGroup: () => invoke<{ ok: boolean }>("end_undo_group"),

    getTimelineState: () => invoke<TimelineResult>("get_timeline_state"),

    // Transport
    setTransport: (payload: { playheadSec?: number; bpm?: number }) =>
        invoke<{ ok: boolean; playhead_sec?: number; bpm?: number }>(
            "set_transport",
            payload.playheadSec,
            payload.bpm,
        ),

    setProjectLength: (projectSec: number) =>
        invoke<TimelineResult>("set_project_length", projectSec),

    // Import
    importAudioItem: (audioPath: string, trackId?: string | null, startSec?: number) =>
        invoke<TimelineResult>("import_audio_item", audioPath, trackId, startSec),

    importAudioBytes: (
        fileName: string,
        base64Data: string,
        trackId?: string | null,
        startSec?: number,
    ) => invoke<TimelineResult>("import_audio_bytes", fileName, base64Data, trackId, startSec),

    getClipSampleAnnotations: (clipId: string) =>
        invoke<ClipSampleAnnotationsResult>("get_clip_sample_annotations", clipId),

    saveClipSampleAnnotations: (
        clipId: string,
        annotations: SampleRegionAnnotation[],
        activeAnnotationIndex: number,
    ) =>
        invoke<SaveClipSampleAnnotationsResult>(
            "save_clip_sample_annotations",
            clipId,
            annotations,
            activeAnnotationIndex,
        ),

    redetectClipSampleAnnotations: (clipId: string) =>
        invoke<ClipSampleAnnotationsResult>("redetect_clip_sample_annotations", clipId),

    openOtoDialog: () =>
        invoke<{ ok: boolean; canceled?: boolean; path?: string; error?: string }>(
            "open_oto_dialog",
        ),

    convertOtoToAnnotations: (otoPath: string) =>
        invoke<{ ok: boolean; result?: OtoConversionResult; error?: string }>(
            "convert_oto_to_annotations",
            otoPath,
        ),

    convertOtoAndRefreshClip: (clipId: string, otoPath: string) =>
        invoke<{
            ok: boolean;
            conversion?: OtoConversionResult;
            clip?: ClipSampleAnnotationsResult;
            error?: string;
        }>("convert_oto_and_refresh_clip", clipId, otoPath),

    // Tracks
    addTrack: (name?: string) => invoke<TimelineResult>("add_track", name),

    addTrackNested: (payload: { name?: string; parentTrackId?: string | null; index?: number }) =>
        invoke<TimelineResult>(
            "add_track",
            payload.name,
            payload.parentTrackId ?? null,
            payload.index,
        ),

    removeTrack: (trackId: string) => invoke<TimelineResult>("remove_track", trackId),

    duplicateTrack: (trackId: string) => invoke<TimelineResult>("duplicate_track", trackId),

    moveTrack: (payload: { trackId: string; targetIndex: number; parentTrackId?: string | null }) =>
        invoke<TimelineResult>(
            "move_track",
            payload.trackId,
            payload.targetIndex,
            payload.parentTrackId ?? null,
        ),

    setTrackState: (payload: {
        trackId: string;
        muted?: boolean;
        solo?: boolean;
        volume?: number;
        composeEnabled?: boolean;
        pitchAnalysisAlgo?: string;
        color?: string;
        name?: string;
    }) =>
        invoke<TimelineResult>(
            "set_track_state",
            payload.trackId,
            payload.muted,
            payload.solo,
            payload.volume,
            payload.composeEnabled,
            payload.pitchAnalysisAlgo,
            payload.color,
            payload.name,
        ),

    selectTrack: (trackId: string) => invoke<TimelineResult>("select_track", trackId),

    getTrackSummary: (trackId?: string) => invoke<TrackSummaryResult>("get_track_summary", trackId),

    // Clips
    addClip: (payload: {
        trackId?: string;
        name?: string;
        startSec?: number;
        lengthSec?: number;
        sourcePath?: string;
    }) =>
        invoke<TimelineResult>(
            "add_clip",
            payload.trackId,
            payload.name,
            payload.startSec,
            payload.lengthSec,
            payload.sourcePath,
        ),

    createClipsBulk: (payload: { templates: ClipTemplate[]; selectCreatedClips?: boolean }) =>
        invoke<TimelineResult>("create_clips_bulk", payload),

    removeClip: (clipId: string) => invoke<TimelineResult>("remove_clip", clipId),

    removeClips: (clipIds: string[]) => invoke<TimelineResult>("remove_clips", clipIds),

    moveClip: (payload: {
        clipId: string;
        startSec: number;
        trackId?: string;
        moveLinkedParams?: boolean;
    }) =>
        invoke<TimelineResult>(
            "move_clip",
            payload.clipId,
            payload.startSec,
            payload.trackId,
            payload.moveLinkedParams,
        ),

    moveClips: (payload: {
        moves: Array<{
            clipId: string;
            startSec: number;
            trackId?: string;
        }>;
        moveLinkedParams?: boolean;
    }) => invoke<TimelineResult>("move_clips", payload.moves, payload.moveLinkedParams),

    getClipLinkedParams: (clipId: string) =>
        invoke<{ ok: boolean; linkedParams?: LinkedParamCurves }>("get_clip_linked_params", clipId),

    applyClipLinkedParams: (payload: { clipId: string; linkedParams: LinkedParamCurves }) =>
        invoke<TimelineResult>("apply_clip_linked_params", payload.clipId, payload.linkedParams),

    setClipState: (payload: {
        clipId: string;
        name?: string;
        startSec?: number;
        lengthSec?: number;
        gain?: number;
        muted?: boolean;
        sourceStartSec?: number;
        sourceEndSec?: number;
        playbackRate?: number;
        reversed?: boolean;
        fadeInSec?: number;
        fadeOutSec?: number;
        fadeInCurve?: string;
        fadeOutCurve?: string;
        color?: string;
        formantMorph?: {
            enabled: boolean;
            targetF1Hz: number;
            targetF2Hz: number;
            strength: number;
        };
        /** 是否创建 undo checkpoint，默认为 true */
        checkpoint?: boolean;
    }) =>
        invoke<TimelineResult>(
            "set_clip_state",
            payload.clipId,
            payload.name,
            payload.startSec,
            payload.lengthSec,
            payload.gain,
            payload.muted,
            payload.sourceStartSec,
            payload.sourceEndSec,
            payload.playbackRate,
            payload.reversed,
            payload.fadeInSec,
            payload.fadeOutSec,
            payload.fadeInCurve,
            payload.fadeOutCurve,
            payload.color,
            payload.formantMorph,
            payload.checkpoint,
        ),

    setClipsStateBulk: (payload: {
        updates: Array<{
            clipId: string;
            gain?: number;
            muted?: boolean;
            fadeInSec?: number;
            fadeOutSec?: number;
        }>;
        checkpoint?: boolean;
    }) => invoke<TimelineResult>("set_clips_state_bulk", payload.updates, payload.checkpoint),

    duplicateClipsBulk: (payload: {
        sourceClipIds: string[];
        deltaSec: number;
        trackMode: Record<string, unknown>;
        copyLinkedParams?: boolean;
        selectCreatedClips?: boolean;
        applyAutoCrossfade?: boolean;
        placeOnSelectedTrack?: boolean;
        renameCopies?: boolean;
    }) => invoke<TimelineResult>("duplicate_clips_bulk", payload),

    replaceClipSource: (payload: {
        clipIds: string[];
        newSourcePath: string;
        replaceSameSource?: boolean;
    }) =>
        invoke<TimelineResult>(
            "replace_clip_source",
            payload.clipIds,
            payload.newSourcePath,
            payload.replaceSameSource,
        ),

    splitClip: (clipId: string, splitSec: number) =>
        invoke<TimelineResult>("split_clip", clipId, splitSec),

    splitClipsAt: (clipIds: string[], splitSec: number) =>
        invoke<TimelineResult>("split_clips_at", clipIds, splitSec),

    glueClips: (clipIds: string[]) => invoke<TimelineResult>("glue_clips", clipIds),

    groupClips: (clipIds: string[]) => invoke<TimelineResult>("group_clips", clipIds),

    ungroupClips: (clipIds: string[]) => invoke<TimelineResult>("ungroup_clips", clipIds),

    toggleGroupDisabled: (groupId: string) =>
        invoke<TimelineResult>("toggle_group_disabled", groupId),

    convertClipsToPitchReference: (clipIds: string[]) =>
        invoke<TimelineResult>("convert_clips_to_pitch_reference", clipIds),

    updatePitchReference: (clipIds: string[]) =>
        invoke<TimelineResult>("update_pitch_reference", clipIds),

    selectClip: (clipId: string | null) => invoke<TimelineResult>("select_clip", clipId),
};
